flang/test/Lower/OpenMP/copyprivate.f90

```diff
@@ -0,0 +1,244 @@
+!CHECK:       %[[SYNC_VAR:.*]]:2 = hlfir.declare %{{.*}} {uniq_name = "_QFtest_typeEt"} : (!fir.ref<!fir.type<_QFtest_typeTty{i:i32,r:f32,a:!fir.array<10xi32>}>>) -> (!fir.ref<!fir.type<_QFtest_typeTty{i:i32,r:f32,a:!fir.array<10xi32>}>>, !fir.ref<!fir.type<_QFtest_typeTty{i:i32,r:f32,a:!fir.array<10xi32>}>>)
+!CHECK:       omp.parallel {
+!CHECK:         omp.single {
+!CHECK:           hlfir.assign %{{.*}}#1 to %[[SYNC_VAR]]#0 temporary_lhs : !fir.ref<!fir.type<_QFtest_typeTty{i:i32,r:f32,a:!fir.array<10xi32>}>>, !fir.ref<!fir.type<_QFtest_typeTty{i:i32,r:f32,a:!fir.array<10xi32>}>>
+!CHECK-NEXT:      omp.terminator
+!CHECK-NEXT:    }
+!CHECK-NEXT:    hlfir.assign %[[SYNC_VAR]]#0 to %{{.*}}#1 temporary_lhs : !fir.ref<!fir.type<_QFtest_typeTty{i:i32,r:f32,a:!fir.array<10xi32>}>>, !fir.ref<!fir.type<_QFtest_typeTty{i:i32,r:f32,a:!fir.array<10xi32>}>>
+!CHECK-NEXT:    omp.barrier
+!CHECK:       }
+subroutine test_type()
+  type ty
+    integer :: i
+    real :: r
+    integer, dimension(10) :: a
+  end type
+
+  integer :: i
+  type(ty) :: t
+
+  t%i = -1
+  t%r = -1.5
+  t%a = -1
+  !$omp parallel firstprivate(t)
+  !$omp single
+  t%i = 42
+  t%r = 3.14
+  do i = 1, 5
+    t%a(i) = i * 10
+  end do
+  !$omp end single copyprivate(t)
+  !$omp end parallel
+end subroutine
+
+!CHECK-LABEL: func @_QPtest_multi
+!CHECK:       %[[I_SYNC_ADDR:.*]] = fir.alloca i32 {bindc_name = "i", pinned, uniq_name = "_QFtest_multiEi"}
+!CHECK:       %[[I_SYNC:.*]]:2 = hlfir.declare %[[I_SYNC_ADDR]] {uniq_name = "_QFtest_multiEi"} : (!fir.ref<i32>) -> (!fir.ref<i32>, !fir.ref<i32>)
+!CHECK:       %[[J_SYNC_ADDR:.*]] = fir.alloca i32 {bindc_name = "j", pinned, uniq_name = "_QFtest_multiEj"}
+!CHECK:       %[[J_SYNC:.*]]:2 = hlfir.declare %[[J_SYNC_ADDR]] {uniq_name = "_QFtest_multiEj"} : (!fir.ref<i32>) -> (!fir.ref<i32>, !fir.ref<i32>)
+!CHECK:       %[[K_SYNC_ADDR:.*]] = fir.alloca i32 {bindc_name = "k", pinned, uniq_name = "_QFtest_multiEk"}
+!CHECK:       %[[K_SYNC:.*]]:2 = hlfir.declare %[[K_SYNC_ADDR]] {uniq_name = "_QFtest_multiEk"} : (!fir.ref<i32>) -> (!fir.ref<i32>, !fir.ref<i32>)
+!CHECK:       omp.parallel {
+!CHECK:         omp.single {
+!CHECK:           %[[I:.*]] = fir.load %{{.*}}#1 : !fir.ref<i32>
+!CHECK:           hlfir.assign %[[I]] to %[[I_SYNC]]#0 temporary_lhs : i32, !fir.ref<i32>
+!CHECK:           %[[J:.*]] = fir.load %{{.*}}#1 : !fir.ref<i32>
+!CHECK:           hlfir.assign %[[J]] to %[[J_SYNC]]#0 temporary_lhs : i32, !fir.ref<i32>
+!CHECK:           %[[K:.*]] = fir.load %{{.*}}#1 : !fir.ref<i32>
+!CHECK:           hlfir.assign %[[K]] to %[[K_SYNC]]#0 temporary_lhs : i32, !fir.ref<i32>
+!CHECK-NEXT:      omp.terminator
+!CHECK-NEXT:    }
+!CHECK-NEXT:    %[[K:.*]] = fir.load %[[K_SYNC]]#0 : !fir.ref<i32>
+!CHECK-NEXT:    omp.barrier
+!CHECK-NEXT:    hlfir.assign %[[K]] to %{{.*}}#1 temporary_lhs : i32, !fir.ref<i32>
+!CHECK-NEXT:    omp.barrier
```

Comment:
In my tests, the private variables were synchronized correctly only with a barrier after the load and another one after the assignment. It should be possible to group the loads and the assignments, however, to use only 2 barriers when there are multiple items in a copyprivate clause.

Reply:
Is this now two per entry in the copyprivate clause?

Reply:
Yes. I'll try to group the loads and assigns, to reduce it to two per copyprivate clause.
Actually, the `omp.barrier` could probably be replaced by a regular memory barrier here, but I couldn't find a `fir` operation for it.

In fact, I'm a bit surprised these barriers are required at all. I expected the end of `single` to synchronize memory among all threads and that no more barriers would be needed after it, for the loads and stores in each thread.
But in practice, the lack of a barrier after an allocatable array load makes its old extent be used in `hlfir.assign`, resulting in a runtime error, and the lack of a barrier after the assign causes old values to be used in a subsequent print statement, for instance.
I have only noticed this issue with allocatable arrays so far.

Reply:
Hmm, maybe an `omp.flush` will do it.

Reply:
@luporl Thanks for the patch. I went through the code and functionality and it looks fine to me, except for the fact that there are too many barriers. Correct me if I am wrong, but the following snippet seems to be the cause of 2 back-to-back barriers after load and assign:

```
        hlfir::Entity r2 = hlfir::loadTrivialScalar(loc, *builder, r);
        if (needBarrier && r2 != r)
          builder->create<mlir::omp::BarrierOp>(loc);
        builder->create<hlfir::AssignOp>(
            loc, r2, l,
            /*isWholeAllocatableAssignment=*/false,
            /*keepLhsLengthInAllocatableAssignment=*/false,
            /*temporary_lhs=*/true);
        if (needBarrier)
          builder->create<mlir::omp::BarrierOp>(loc);
```

However, I am not sure whether reducing this to one barrier may still cause one barrier per copyprivate list operation (since `copyVar` is called on each symbol). As such, does it make sense to factor out generation of barrier too from `copyVar` and explicitly generate a barrier at the end? Or maybe `needBarrier=false` for all-but-last variable in the copyprivate list? 

Rest all LGTM!

Reply:
@NimishMishra Thanks for the review! That's right, this is the part that inserts 2 barriers.
The problem is that, if I reduce it to one barrier, I get incorrect results when using copyprivate with allocatable arrays (tested on AArch64). It should be possible to refactor the code to generate all loads of the copyprivate list, then a barrier, then all assigns and then another barrier.

But, to be honest, I still don't understand why it only works with 2 barriers. I will try to investigate this further, to see if there is a better solution.

Reply:
I found out the issue with allocatable arrays was actually that the code that performs the deallocation of the sync var was ending up in the parallel region. The barriers were just preventing the sync var of being deallocated while still being used.

Reply:
The new implementation is going to use the `__kmpc_copyprivate` runtime function, which avoids the need to manually insert barriers.